File: backend/internal/utils/wordcount.go
package utils

import (
	"strings"
	"unicode"
)

func CountStoryUnits(text string) int {
	count := 0
	inLatinToken := false
	for _, r := range strings.TrimSpace(text) {
		switch {
		case unicode.Is(unicode.Han, r) || unicode.Is(unicode.Hiragana, r) || unicode.Is(unicode.Katakana, r) || unicode.Is(unicode.Hangul, r):
			if inLatinToken {
				inLatinToken = false
			}
			count++
		case unicode.IsLetter(r) || unicode.IsDigit(r):
			if !inLatinToken {
				count++
				inLatinToken = true
			}
		default:
			inLatinToken = false
		}
	}
	return count
}

File: frontend/src/utils/wordCount.ts
export function countStoryUnits(text: string): number {
  let count = 0;
  let inLatin = false;
  for (const char of text.trim()) {
    if (/[\p{Script=Han}\p{Script=Hiragana}\p{Script=Katakana}\p{Script=Hangul}]/u.test(char)) {
      inLatin = false;
      count += 1;
    } else if (/[\p{L}\p{N}]/u.test(char)) {
      if (!inLatin) count += 1;
      inLatin = true;
    } else {
      inLatin = false;
    }
  }
  return count;
}
